[Lens] Xy scale type

flash1293 opened the pull request:
Pass the type of a table column to the expression render function of xy charts and use this information to specify the correct scale type for elastic-charts.

It also extends the operation meta data to also carry a flag for a histogram column and information about the scale. If the histogram flag is set, the histogram mode of elastic charts is used. In a further extension this could also be configurable by the user.

The `isHistogram` flag and the `scale` on the operation can also be used for suggestions later - e.g. a pie chart makes sense for a bucketed dimension that is not a histogram (e.g. ranges or filters) regardless of data type, but not really for a histogram dimension (here you should use a cartesian chart). A gauge chart would make sense for a ratio scale, but nor for an interval scale. A terms operation of a number field is of type number, but should be displayed as an ordinal scale instead of a linear scale. Information like this can be used to fine tune the suggestions and to boost scores for certain configurations.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9774/)

wylieconlon commented on x-pack/legacy/plugins/lens/public/xy_visualization_plugin/xy_expression.tsx:
```diff
@@ -100,6 +100,33 @@ function getIconForSeriesType(seriesType: SeriesType): IconType {
+  } else {
+    const minDate = ofAllXValues(Math.min);
+    const maxDate = ofAllXValues(Math.max);
+    return niceTimeFormatter([minDate, maxDate]);
```
This logic doesn't really make sense to me- the extent of the data should be unrelated to the formatting. I would expect that based on the interval in the data source, the formatter would change precisions- regardless of how much data is returned

flash1293 replied:
You are right, this is more like a rule of thumb which takes advantage of the fact that the density of data points shown on the screen doesn't change that much - because of that, the interval and the extent of the data correlates.

The rule behind `niceTimeFormatter` is
```
  if (days > 30) {
    return 'yyyy-MM-dd';
  }
  if (days > 7 && days <= 30) {
    return 'MMMM dd';
  }
  if (days > 1 && days <= 7) {
    return 'MM-dd HH:mm';
  }
  return 'HH:mm:ss';
```

Which I think is good enough as a temporary solution for beta. In the long run this should be configurable in the chart itself.

flash1293 commented:
> This is working well without touching the datasource APIs- seems like that's the next step?
Right, we could add something to the column meta data. But I'm not sure whether we need that for our beta launch.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10023/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11955/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11986/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12095/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12108/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12148/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12172/)

wylieconlon commented on x-pack/legacy/plugins/lens/public/types.ts:
```diff
@@ -158,7 +158,8 @@ export interface OperationMetadata {
   isBucketed: boolean;
-
+  isHistogram?: boolean;
+  scale?: 'ordinal' | 'interval' | 'ratio';
```
I'm not sure I understand the need for these additions- can you explain?

flash1293 replied:
The histogram information is used to enable histogram mode if that makes sense for the operation. The scale is not necessary for the currently supported operations, but will come in handy later. E.g. if we do a terms of a number field, the data type will be number, but we shouldn't use a linear scale in that case. This field makes it possible for the visualization to react to that differences (e.g. providing different suggestions). See also the PR description.

wylieconlon commented on x-pack/legacy/plugins/lens/public/xy_visualization_plugin/to_expression.ts:
```diff
@@ -55,12 +56,55 @@ export const toExpression = (state: State, frame: FramePublicAPI): Ast | null =>
+      return ScaleType.Time;
+    default:
+      return ScaleType.Linear;
+  }
```
I don't see this using `metadata.isHistogram`?

flash1293 replied:
It's passed in as a separate property in line 169. I tried to stick closely to the elastic-charts API when possible.

flash1293 commented:
@wylieconlon 
> I think I'm not seeing the goal of this PR, is there anything user-visible here?

The only user facing change is that the chart is using histogram mode (https://elastic.github.io/elastic-charts/?path=/story/bar-chart--test-histogram-mode-linear) for date histogram which makes the bar chart an actual histogram (better tick placement and width of the bars matches covered range of x axis)

> I was seeing the date formatting work as expected with Date Histogram + count, but formatting for filter ratio is not working as I was expecting:

The argument to include format hints to esaggs was missing in the separate expression string used for filter ratio, I added it here.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12471/)

flash1293 commented:
@wylieconlon I thought about the metadata again and I think `isHistogram` is redundant - it's just `isBucketed && scale !== 'ordinal'`, because each bucketed (or binned) dimension that isn't ordinal can serve that purpose - I removed it and changed the logic in the xy expression layer accordingly. We should try to keep the operation meta data as simple as possible, but still keep it on an abstract layer to avoid coupling to index patterns and ES aggregations too much.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12563/)
